### PR TITLE
Nationalism AI; changes for encouraging more nationalism to successfully suceed

### PR DESCRIPTION
**Pull request by ma44**
Big list of changes
- Rebel AI can now properly form pan nationalist factions, although nationalists still appear to be broken, this should drastically increase the chance of an Italy appearing alongside the occasional Gutter crown for the germans
- Nations will try to unally and desphere countries they have wargoals against; stops certain situations like USA sphering mexico causing them to not be able to manifest a destiny
- Modified some tech weights, effective outcome should be a much higher focus on national focus techs, culture and army
- Redid the sphere AI; now tries to not sphere countries they don't have a neighboring sphere or a coastal state to access. Countries will now hyperfocus on sphering countries within their culture union
- AI will now intervene in wars even if they're already in a war if it means they can protect a non-GP culture union; this could use some military strength checks, but this is mainly to ensure the USA will keep Texas safe from the Mexicans and not be pseudo locked out from manifest destiny
- Wars that extend for > 6 years and have minor amounts of war score will now be peaced out easily; stops forever stalemate wars particularly with Europeans doing "cut down to size" wars against asian countries. 
- Countries with very high war exhaustion will peace out after 1 year in war; countries locked in rebel death spiral will try to escape the situation, especially when the aggressor cannot occupy the lands because of rebels
- GPs will now look globally to figure out where they can use their wargoals; helps Prussia decide in attacking Denmark and would allow for other free wargoals to be used like 'pay debts' and 'cut down to size'
- Undid the rebel size reduction, rebels winning a revolution will now negate militancy; this should help nations not ener a permanent death spiral just because the population is permanently unhappy

### CODE REVIEW

File: src/ai/ai.cpp [hunk 1]
```diff
@@ -166,8 +171,12 @@ void prune_alliances(sys::state& state) {
 		if(!n.get_is_player_controlled() && !n.get_ai_is_threatened() && !(n.get_overlord_as_subject().get_ruler())) {
 			prune_targets.clear();
 			for(auto dr : n.get_diplomatic_relation()) {
+				auto other = dr.get_related_nations(0) != n ? dr.get_related_nations(0) : dr.get_related_nations(1);
```
Comment (schombert) [inline src/ai/ai.cpp:174]: I don't think there is any point putting this check outside of the are allied check. The prune list is only used to remove existing alliances, and checking for possible cbs against nations you aren't allied to here would be kind of a waste if there is nothing to prune.

File: src/ai/ai.cpp [hunk 2]
```diff
@@ -270,6 +279,11 @@ bool ai_will_accept_alliance(sys::state& state, dcon::nation_id target, dcon::na
 	if(!ai_is_close_enough(state, target, from))
 		return false;
 
+	//We'd rather want to kill them rather than ally
+	if(military::can_use_cb_against(state, from, target)) {
+		return false;
```
Comment (schombert) [inline src/ai/ai.cpp:284]: This will need a corresponding line added to the offer alliance button in the diplomacy window explaining why the ai won't accept

File: src/ai/ai.cpp [hunk 3]
```diff
@@ -385,30 +399,30 @@ void initialize_ai_tech_weights(sys::state& state) {
 	for(auto t : state.world.in_technology) {
 		float base = 1000.0f;
 		if(state.culture_definitions.tech_folders[t.get_folder_index()].category == culture::tech_category::army)
-			base *= 1.5f;
+			base *= 3.0f;
 
 		if(t.get_increase_building(economy::province_building_type::naval_base))
-			base *= 1.1f;
+			base *= 9.0f;
```
Comment (schombert) [inline src/ai/ai.cpp:405]: really, *nine* times more weight to increasing naval base level? Nothing should have more weight than the weight given to techs that increase the research point modifier. I should also point out that only the relative values matter. Increasing all the weights won't really matter; only the relative weightings matter in most cases

File: src/ai/ai.cpp [hunk 4]
```diff
@@ -532,6 +583,9 @@ void perform_influence_actions(sys::state& state) {
 				command::execute_remove_from_sphere(state, gprl.get_great_power(), gprl.get_influence_target(), gprl.get_influence_target().get_in_sphere_of());
 			} else if(state.defines.addtosphere_influence_cost <= gprl.get_influence() && !current_sphere && clevel == nations::influence::level_friendly) {
 				command::execute_add_to_sphere(state, gprl.get_great_power(), gprl.get_influence_target());
+			//De-sphere countries we have wargoals against
+			} else if(military::can_use_cb_against(state, gprl.get_great_power(), gprl.get_influence_target()) && state.defines.removefromsphere_influence_cost <= gprl.get_influence() && current_sphere && clevel == nations::influence::level_friendly) {
+				command::execute_remove_from_sphere(state, gprl.get_great_power(), gprl.get_influence_target(), gprl.get_influence_target().get_in_sphere_of());
```
Comment (schombert) [inline src/ai/ai.cpp:588]: This needs to check for whether the nation has the infamy room. We wouldn't want a nation pushing themselves over the infamy cap for this

File: src/ai/ai.cpp [hunk 5]
```diff
@@ -1833,11 +1887,23 @@ bool will_accept_crisis_peace_offer(sys::state& state, dcon::nation_id to, dcon:
 
 void update_war_intervention(sys::state& state) {
 	for(auto& gp : state.great_nations) {
-		if(state.world.nation_get_is_player_controlled(gp.nation) == false && state.world.nation_get_is_at_war(gp.nation) == false) {
```
Comment (schombert) [inline src/ai/ai.cpp:1836]: Are you sure removing this is a good idea? I don't think that the AI should generally be jumping into more wars if it is already at war

File: src/ai/ai.cpp [hunk 6]
```diff
@@ -1833,11 +1887,23 @@ bool will_accept_crisis_peace_offer(sys::state& state, dcon::nation_id to, dcon:
...
+		if(state.world.nation_get_is_player_controlled(gp.nation) == false){
 			bool as_attacker = false;
 			dcon::war_id intervention_target;
 			[&]() {
 				for(auto w : state.world.in_war) {
+					//GPs will try to intervene in wars to protect smaller nations in the same cultural union
+					if(command::can_intervene_in_war(state, gp.nation, w, false)) {
+						for(auto par : w.get_war_participant()) {
```
Comment (schombert) [inline src/ai/ai.cpp:1897]: instead of looping over everyone in the war, just check the primary defender -- this will prevent the US jumping into any war that texas gets involved in as a defensive ally, for an example of the weird things that could happen

File: src/ai/ai.cpp [hunk 7]
```diff
@@ -2735,6 +2800,14 @@ bool will_accept_peace_offer(sys::state& state, dcon::nation_id n, dcon::nation_
 		if(war_duration < 365) {
 			return false;
 		}
+		//The country is ruined and it's been a bit into the war, we'll recover later
+		if(state.world.nation_get_war_exhaustion(n) > 80) {
+			return true;
+		}
+		//Stalemated war, peace out
+		if(war_duration > (365 * 6) && (overall_score <= 10 && overall_score >= -10)) {
```
Comment (schombert) [inline src/ai/ai.cpp:2808]: maybe only if the offer is a white peace? Otherwise this will let you take 100% warscore in a stalemated war

File: src/ai/ai.cpp [hunk 8]
```diff
@@ -2735,6 +2800,14 @@ bool will_accept_peace_offer(sys::state& state, dcon::nation_id n, dcon::nation_
 		if(war_duration < 365) {
 			return false;
 		}
+		//The country is ruined and it's been a bit into the war, we'll recover later
+		if(state.world.nation_get_war_exhaustion(n) > 80) {
```
Comment (schombert) [inline src/ai/ai.cpp:2804]: these changes will also need to be reflected in the function that calculates the same thing with a bunch of extra parameters (this is used for evaluating a peace offer possibility without actually making it, mostly for the ui)

File: src/culture/rebels.cpp [hunk 9]
```diff
@@ -389,26 +402,13 @@ bool pop_is_compatible_with_rebel_faction(sys::state& state, dcon::pop_id p, dco
...
-		return true;
-	}
 	return true;
 }
 
 bool pop_is_compatible_with_rebel_type(sys::state& state, dcon::pop_id p, dcon::rebel_type_id t) {
 	auto fac = fatten(state.world, t);
 	auto pop = fatten(state.world, p);
 
-	if(fac.get_ideology() && fac.get_ideology() != pop.get_dominant_ideology())
```
Comment (schombert) [inline src/culture/rebels.cpp:410]: why did you remove this check? This prevents, say, pops that are mostly reactionary from joining communist rebels, etc

Comment (schombert) [reply src/culture/rebels.cpp:410]: oh, I see, it just got moved

**Comment by schombert** [comment]
Something weird seems to be going on here. This PR says that it was merged, but I don't recall doing that (and didn't intended to before the issues raised above were addressed). Nor does it appear to be actually merged (its changes don't appear to be in the source files). Can anyone clue me into what is going on?

**Comment by ma44** [comment]
Mystery solved
